flink-state-backends/flink-statebackend-heap-spillable/src/main/java/org/apache/flink/runtime/state/heap/space/Allocator.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state.heap.space;
+
+import java.io.Closeable;
+
+/**
+ * Implementations are responsible for allocate space.
+ */
+public interface Allocator extends Closeable {
+
+	/**
+	 * Allocate space with the given size.
+	 *
+	 * @param size size of space to allocate.
+	 * @return address of the allocated space, or -1 when allocation is failed.
+	 */
+	long allocate(int size);
```

Comment:
We will check against `-1`, see `SkipListUtils.NIL_VALUE_POINTER`.

We don't throw exception here because the definition of methods in `StateMap` don't throw exceptions, so we will need to handle the exception somewhere if we throw any in lower level functions, then no big difference from handling `-1`.

Reply:
I don't understand it. If the space allocation fails, then we should also fail the operation right? How exactly will this be caught? It looks as if we pass the return value of `allocate` unconditionally to other methods. Are you saying that these methods will simply swallow this value or fail at a different place?

Reply:
I mean we're checking the return value, not `-1` literally but using a constant `SkipListUtils.NIL_VALUE_POINTER` to represent it.

Reply:
Ok in order to clarify my question: In `CopyOnWriteSkipListStateMap#putNode:369` how do we make sure that we obtain a valid node pointer? It looks to me as if we simply forward the pointer and let it fail somewhere else.

Reply:
ok, got the question, and yes we should better handle this part, will resolve this in new commit.